### ISSUE TITLE
segfault in gg_ttgg runTest.exe in my 'june24' WIP

### ISSUE THREAD

**Issue by valassi** [madgraph5/madgraph4gpu#917]
There are several segfaults in gg_ttgg runTest.exe in my 'june24' WIP

This is part of the 24 failures here

Reproducible locally. In debug mode
```
Program received signal SIGSEGV, Segmentation fault.
0x0000000000424611 in mg5amcCpu::sigmaKin (allmomenta=0x519340, allcouplings=0x52a580, allrndhel=0x525c40, allrndcol=0x5264c0, 
    allMEs=0x5275c0, allChannelIds=0x518ec0, allNumerators=0x52d600, allDenominators=0x52de80, allselhel=0x527e00, allselcol=0x528240, 
    nevt=256) at CPPProcess.cc:3327
3327              if( mgOnGpu::icolamp[iconfig - 1][icolC] ) targetamp[icolC] += jamp2_sv[icolC];
Missing separate debuginfos, use: dnf debuginfo-install glibc-2.34-60.el9.x86_64 libgcc-11.3.1-4.3.el9.alma.x86_64 libstdc++-11.3.1-4.3.el9.alma.x86_64
(gdb) where
#0  0x0000000000424611 in mg5amcCpu::sigmaKin (allmomenta=0x519340, allcouplings=0x52a580, allrndhel=0x525c40, allrndcol=0x5264c0, 
    allMEs=0x5275c0, allChannelIds=0x518ec0, allNumerators=0x52d600, allDenominators=0x52de80, allselhel=0x527e00, allselcol=0x528240, 
    nevt=256) at CPPProcess.cc:3327
#1  0x0000000000437502 in mg5amcCpu::MatrixElementKernelHost::computeMatrixElements (this=0x528750, useChannelIds=true)
    at MatrixElementKernels.cc:221
#2  0x0000000000470346 in CPUTest::runSigmaKin (this=0x4fcd60 <driver2>, iiter=0) at runTest.cc:131
#3  0x000000000046c55f in MadgraphTest::CompareMomentaAndME (this=0x4fd248 <mgTest2>, googleTest=...)
    at /data/avalassi/GPU2023/madgraph4gpuX/epochX/cudacpp/gg_ttgg.mad/SubProcesses/P1_gg_ttxgg/MadgraphTest.h:218
#4  0x000000000046dcff in SIGMA_SM_GG_TTXGG_CPU_MULTICHANNEL_compareMomAndME_Test::TestBody (this=0x54cbd0) at runTest.cc:356
#5  0x00000000004aa07c in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void> (object=0x54cbd0, 
    method=&virtual testing::Test::TestBody(), location=0x4c5df3 "the test body")
    at /data/avalassi/GPU2023/madgraph4gpuX/test/googletest/googletest/src/gtest.cc:2607
#6  0x00000000004a3455 in testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void> (object=0x54cbd0, 
    method=&virtual testing::Test::TestBody(), location=0x4c5df3 "the test body")
    at /data/avalassi/GPU2023/madgraph4gpuX/test/googletest/googletest/src/gtest.cc:2643
#7  0x000000000047b948 in testing::Test::Run (this=0x54cbd0)
    at /data/avalassi/GPU2023/madgraph4gpuX/test/googletest/googletest/src/gtest.cc:2682
#8  0x000000000047c299 in testing::TestInfo::Run (this=0x54ce70)
    at /data/avalassi/GPU2023/madgraph4gpuX/test/googletest/googletest/src/gtest.cc:2861
#9  0x000000000047caa9 in testing::TestSuite::Run (this=0x54cf90)
    at /data/avalassi/GPU2023/madgraph4gpuX/test/googletest/googletest/src/gtest.cc:3015
#10 0x000000000048c002 in testing::internal::UnitTestImpl::RunAllTests (this=0x510050)
    at /data/avalassi/GPU2023/madgraph4gpuX/test/googletest/googletest/src/gtest.cc:5855
#11 0x00000000004ab1cd in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x510050, 
    method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl * const)) 0x48bc3e <testing::internal::UnitTestImpl::RunAllTests()>, location=0x4c6858 "auxiliary test code (environments or event listeners)")
    at /data/avalassi/GPU2023/madgraph4gpuX/test/googletest/googletest/src/gtest.cc:2607
#12 0x00000000004a43cd in testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x510050, 
    method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl * const)) 0x48bc3e <testing::internal::UnitTestImpl::RunAllTests()>, location=0x4c6858 "auxiliary test code (environments or event listeners)")
    at /data/avalassi/GPU2023/madgraph4gpuX/test/googletest/googletest/src/gtest.cc:2643
#13 0x000000000048a9a8 in testing::UnitTest::Run (this=0x4fd480 <testing::UnitTest::GetInstance()::instance>)
    at /data/avalassi/GPU2023/madgraph4gpuX/test/googletest/googletest/src/gtest.cc:5438
#14 0x000000000045d3c8 in RUN_ALL_TESTS () at ../../../../../test/googletest/install_gcc11.3.1/include/gtest/gtest.h:2490
#15 0x000000000045cfb0 in main (argc=1, argv=0x7fffffffda58) at testxxx.cc:453
(gdb) l
3322            {
3323              if( icolC == 0 )
3324                targetamp[icolC] = fptype_sv{ 0 };
3325              else
3326                targetamp[icolC] = targetamp[icolC - 1];
3327              if( mgOnGpu::icolamp[iconfig - 1][icolC] ) targetamp[icolC] += jamp2_sv[icolC];
3328            }
3329    #if defined MGONGPU_CPPSIMD and defined MGONGPU_FPTYPE_DOUBLE and defined MGONGPU_FPTYPE2_FLOAT
3330            fptype_sv targetamp2[ncolor] = { 0 };
3331            for( int icolC = 0; icolC < ncolor; icolC++ )
(gdb) p iconfig
$1 = 4294967295
(gdb) p channelId
$2 = 1

```

This is a cppnone build

**Comment by valassi** [comment]
Well well. The segfault makes sense. This is channelId=1 and this has no associated iconfig...
```
  // Map channel to iconfig (e.g. "iconfig = channel2iconfig[channelId - 1]": input index uses C indexing, output index uses F indexing)
  // Note: iconfig=-1 indicates channels/diagrams with no associated iconfig for single-diagram enhancement in the MadEvent sampling algorithm (presence of 4-point interaction?)
  // This array has N_diagrams elements, but only N_config <= N_diagrams valid values (iconfig>0)
  __device__ constexpr int channel2iconfig[113] = { // note: a trailing comma in the initializer list is allowed
      -1, // CHANNEL_ID=1    i.e. DIAGRAM=1    --> ICONFIG=-1 (diagram with no associated iconfig for single-diagram enhancement)
       1, // CHANNEL_ID=2    i.e. DIAGRAM=2    --> ICONFIG=1
       2, // CHANNEL_ID=3    i.e. DIAGRAM=3    --> ICONFIG=2
```

I need to do two things:
- first, add a sanity check with an assert to prevent a segfault (better an internal error assert than a segfault)
- second, configure my multichannel test so that it uses only channels that have an associated iconfig...

En passant, this is probably the reason why I got these two disturbing errors
- here 7e12a792b56d394fce168cb4e67fbaf0ddeeb91e
- and here 586c5448de7d981fa09701afbed08c1f4ae228d0

**Comment by valassi** [comment]
Ok the segfault is gone, here comes the assert
```
diff --git a/epochX/cudacpp/gg_ttgg.mad/SubProcesses/P1_gg_ttxgg/CPPProcess.cc b/epochX/cudacpp/gg_ttgg.mad/SubProcesses/P1_gg_ttxgg/CPPProcess.cc
index b358eec84..504008914 100644
--- a/epochX/cudacpp/gg_ttgg.mad/SubProcesses/P1_gg_ttxgg/CPPProcess.cc
+++ b/epochX/cudacpp/gg_ttgg.mad/SubProcesses/P1_gg_ttxgg/CPPProcess.cc
@@ -3173,7 +3173,14 @@ namespace mg5amcCpu
     if( channelId != 0 ) // no event-by-event choice of color if channelId == 0 (fix FPE #783)
     {
       // NB (see #877): in the array channel2iconfig, the input index uses C indexing (channelId -1), the output index uses F indexing (iconfig)
-      const unsigned int iconfig = mgOnGpu::channel2iconfig[channelId - 1]; // map N_diagrams to N_config <= N_diagrams configs (fix LHE color mismatch #856: see also #826, #852, #853)
+      // NB (see #917): mgOnGpu::channel2iconfig returns an int (which may be -1), not an unsigned int!
+      const int iconfig = mgOnGpu::channel2iconfig[channelId - 1]; // map N_diagrams to N_config <= N_diagrams configs (fix LHE color mismatch #856: see also #826, #852, #853)
+      if( iconfig <= 0 )
+      {
+        std::cout << "INTERNAL ERROR! Cannot choose an event-by-event random color for channelId=" << channelId
+                  << " which has no associated iconfig for single-diagram enhancement" << std::endl;
+        assert( iconfig > 0 ); // SANITY CHECK #917
+      }
       fptype targetamp[ncolor] = { 0 };
       // NB (see #877): explicitly use 'icolC' rather than 'icol' to indicate that icolC uses C indexing in [0, N_colors-1]
       for( int icolC = 0; icolC < ncolor; icolC++ )
@@ -3315,7 +3322,14 @@ namespace mg5amcCpu
       if( channelId != 0 ) // no event-by-event choice of color if channelId == 0 (fix FPE #783)
       {
         // NB (see #877): in the array channel2iconfig, the input index uses C indexing (channelId -1), the output index uses F indexing (iconfig)
-        const unsigned int iconfig = mgOnGpu::channel2iconfig[channelId - 1]; // map N_diagrams to N_config <= N_diagrams configs (fix LHE color mismatch #856: see also #826, #852, #853)
+        // NB (see #917): mgOnGpu::channel2iconfig returns an int (which may be -1), not an unsigned int!
+        const int iconfig = mgOnGpu::channel2iconfig[channelId - 1]; // map N_diagrams to N_config <= N_diagrams configs (fix LHE color mismatch #856: see also #826, #852, #853)
+        if( iconfig <= 0 )
+        {
+          std::cout << "INTERNAL ERROR! Cannot choose an event-by-event random color for channelId=" << channelId
+                    << " which has no associated iconfig for single-diagram enhancement" << std::endl;
+          assert( iconfig > 0 ); // SANITY CHECK #917
+        }
         fptype_sv targetamp[ncolor] = { 0 };
         // NB (see #877): explicitly use 'icolC' rather than 'icol' to indicate that icolC uses C indexing in [0, N_colors-1]
         for( int icolC = 0; icolC < ncolor; icolC++ )
```

and
```
[----------] 1 test from SIGMA_SM_GG_TTXGG_CPU_MULTICHANNEL
[ RUN      ] SIGMA_SM_GG_TTXGG_CPU_MULTICHANNEL.compareMomAndME
INFO: Opening reference file ../../test/ref/dump_CPUTest.Sigma_sm_gg_ttxgg.txt2
INTERNAL ERROR! Cannot choose an event-by-event random color for channelId=1 which has no associated iconfig for single-diagram enhancement
runTest_cpp.exe: CPPProcess.cc:3331: void mg5amcCpu::sigmaKin(const fptype*, const fptype*, const fptype*, const fptype*, mgOnGpu::fptype*, const unsigned int*, mgOnGpu::fptype*, mgOnGpu::fptype*, int*, int*, int): Assertion `iconfig > 0' failed.
Aborted (core dumped)
```

**Comment by valassi** [comment]
> * configure my multichannel test so that it uses only channels that have an associated iconfig...

This was the main issue, fixed here
https://github.com/valassi/madgraph4gpu/commit/78dd1d16759dfed65518b086002eeb746fe5928e
```
    static_assert( nevt % warpSize == 0, "ERROR! nevt should be a multiple of warpSize" );
    constexpr unsigned int nWarp = nevt / warpSize;
    for( unsigned int iWarp = 0; iWarp < nWarp; ++iWarp )
    {
      //const unsigned int channelId = 1 + ( iWarp + iiter * nWarp ) % CPPProcess::ndiagrams; // bug #917
      const int iconfig = 1 + ( iWarp + iiter * nWarp ) % mgOnGpu::nconfigSDE;
      unsigned int channelId = 0;
      for( unsigned int idiagram = 1; idiagram < CPPProcess::ndiagrams; idiagram++ )
      {
        if( mgOnGpu::channel2iconfig[idiagram] == iconfig )
        {
          channelId = idiagram+1; // fix #917 (NB add +1 because channelId uses F indexing)
          break;
        }
      }
      assert( channelId > 0 ); // sanity check that the channelId for the given iconfig was found
      std::cout << "CUDA_CPU_TestBase::setChannelIds: iWarp=" << iWarp << ", iconfig=" << iconfig << ", channelId=" << channelId << std::endl;
```

Done in 882, closing
